### ISSUE TITLE
Can't even install - Java Runtime Error

### ISSUE THREAD

**Issue by fasafan** [triplea-game/triplea#2668]
Trying to install TripleA 1.9.0.0.7621 Windows 32 bit.

Windows XP, SP2 (this machine is used for legacy gaming mostly in DOSBox; I don't plan on online play.)

Install ends almost as soon as it begins with an "Error downloading the Java Runtime Environment.  Please check your internet connection and start setup again" while at the "...downloading the Java Runtime Environment" step.

Computer is connected to the internet.

Java version is 7 Update 79.

Any help would be greatly appreciated.  I was playing on this machine three days ago or so, but I fouled the registry as was forced to reinstall Windows XP.

Thanks for any help.

Lee

**Comment by ssoloff** [comment]
I tried installing 1.9.0.0.7621 on a Windows x86 box without Java 8 installed, and install4j successfully downloaded and used the bundled Java 8 JRE we provide in the triplea-game/assets repo, so this doesn't appear to be a general issue.

@fasafan There's a couple things you can do here.

First, the installer generates a log that we can take a look at to see why it's failing to download the Java 8 JRE on your machine.  It's been a while since I've used XP, but I think this file should be in the folder _C:\Documents and Settings\fasafan\Local Settings\Temp_, where _fasafan_ should be replaced with your Windows username.  The file name will look something like _i4j\_nlog\_#.log_, where _#_ could be 1, 2, 3, etc.  Please attach that file to this issue.  If there are multiple log files whose name matches that format, just attach the latest one.

Second, you can try to avoid the issue by downloading the Java 8 JRE from [Oracle](http://www.oracle.com/technetwork/java/javase/downloads/jre8-downloads-2133155.html) and installing it yourself.  This will cause the TripleA installer to bypass trying to download the Java 8 JRE.  Note that Java 8 isn't officially supported on XP, and I believe you'll see a message during installation of the JRE to that effect.

**Comment by ssoloff** [comment]
@fasafan One other question...  If you click this link:

https://raw.githubusercontent.com/triplea-game/assets/master/install4j/windows-x86-1.8.0_144.tar.gz 

from a browser on your XP machine, do you get a security warning saying something like "the certificate is invalid, do you want to continue?"

**Comment by fasafan** [comment]
Hello!  Thanks for giving me where that .log file was located.  I looked for it, but couldn't find it.  
I'll copy it below since I'm not sure how to attach a file.
I can open that link you posted.  In Firefox, it wants me to open or save the file...?
~~~~~~~~~~~~~~~
Started executable C:\Documents and Settings\Owner\My Documents\Downloads\TripleA_1.9.0.0.7621_windows-32bit.exe at Mon Dec 04 00:19:15 2017

[0:0] init file name C:\Documents and Settings\Owner\My Documents\Downloads\TripleA_1.9.0.0.7621_windows-32bit.exe C:\Documents and Settings\Owner\My Documents\Downloads\TripleA_1.9.0.0.7621_windows-32bit.exe 93 1309
[0:0] number of sections: 5
[0:0] size of optional headers: 224
[0:0] resSectionTableStart: 648
[0:0] rawDataSize: 15360, rawDataOffset: 813568
[0:0] sun.locale.formatasdefault is false
[0:0] language/country is en_US
[0:0] change working directory to C:\Documents and Settings\Owner\My Documents\Downloads
[0:0] single instance mode
[0:0] semaphore name Local\c:_documents_and_settings_owner_my_documents_downloads_triplea_1.9.0.0.7621_windows-32bit.exe, code 0, value 00000044
[0:0] Init done
[0:20] Starting work
[0:20] number of sections: 5
[0:20] size of optional headers: 224
[0:20] resSectionTableStart: 648
[0:20] rawDataSize: 15360, rawDataOffset: 813568
[0:20] starting at 828928
[0:20] verifying integrity length 10247163
[0:721] checksums: 1524375560 1524375560 1524375560 1524375560
[0:721] tempPath is C:\DOCUME~1\Owner\LOCALS~1\Temp\
[0:721] tempFile is C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp
[0:721] Temp dir3: C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356
[0:721] native lib dir is ${launcher:sys.launcherTempDirectory}\user
[0:721] expanding2 ${launcher:sys.launcherTempDirectory} to C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356 in C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\user.
[0:721] Setting PATH (native libs) C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\user;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem
[0:721] extracting files
[0:721] Included files: 15
[0:721] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4jruntime.jar.pack (size: 735640)
[0:731] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4jparams.conf (size: 198517)
[0:731] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\MessagesDefault (size: 33668)
[0:731] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\user.jar.pack (size: 4951)
[0:731] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_0_1sxis30.utf8 (size: 33668)
[0:731] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_1_1sxis30_1r5xq07.png (size: 860)
[0:731] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\stats.properties (size: 5320)
[0:741] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_2_1sxis30_1e4h48z.png (size: 1981)
[0:741] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_4_1sxis30_1wwgbbl.png (size: 5319)
[0:741] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_7_1sxis30_18gg8kx@2x.png (size: 15419)
[0:741] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\installer.ico (size: 476334)
[0:741] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_6_1sxis30_8of556.png (size: 37150)
[0:741] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_3_1sxis30_tdx38l.png (size: 3739)
[0:741] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_7_1sxis30_18gg8kx.png (size: 9202)
[0:741] extracting C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_5_1sxis30_1vuodf0.png (size: 14152)
[0:741] extracting files done
[0:741] Current working dir: C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356
[0:741] Full JRE path: C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\jre.tar.gz
[0:741] No JRE included
[0:741] LoadDLL (0, (null), 0)
[0:741] Searching for a JVM
[0:741] MinVersion: 1.8, MaxVersion: 1.8
[0:751] Testing location  (type Y)
[0:751] Testing location JAVA_HOME (type E)
[0:751] Testing location JDK_HOME (type E)
[0:751] checking app id allinstdirs5251-3669-9623-1649 (0, 80000001)
[0:751] could not open installation node 2
[0:751] checking app id allinstdirs5251-3669-9623-1649 (512, 80000001)
[0:751] could not open installation node 2
[0:751] checking app id allinstdirs5251-3669-9623-1649 (0, 80000002)
[0:751] could not open installation node 2
[0:751] checking app id allinstdirs5251-3669-9623-1649 (512, 80000002)
[0:751] could not open installation node 2
[0:751] Search sequence finished
[0:751] ERROR: No JVM found
[0:751] Download location: https://raw.githubusercontent.com/triplea-game/assets/master/install4j/windows-x86-1.8.0_144.tar.gz
[0:751] Downloading JRE
[1:773] ERROR: InternetOpenUrl failed with error code -2146869244
[2:774] ERROR: Retry JRE download
[2:965] ERROR: InternetOpenUrl failed with error code -2146869244
[3:966] ERROR: Retry JRE download
[6:960] ERROR: InternetOpenUrl failed with error code -2146869244
[7:962] ERROR: Retry JRE download
[10:305] emptying C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4jparams.conf
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4jruntime.jar.pack
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_0_1sxis30.utf8
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_1_1sxis30_1r5xq07.png
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_2_1sxis30_1e4h48z.png
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_3_1sxis30_tdx38l.png
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_4_1sxis30_1wwgbbl.png
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_5_1sxis30_1vuodf0.png
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_6_1sxis30_8of556.png
[10:305] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_7_1sxis30_18gg8kx.png
[10:315] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\i4j_extf_7_1sxis30_18gg8kx@2x.png
[10:315] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\installer.ico
[10:315] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\MessagesDefault
[10:315] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\stats.properties
[10:315] attempting to delete C:\DOCUME~1\Owner\LOCALS~1\Temp\e4j1B.tmp_dir1512368356\user.jar.pack

**Comment by fasafan** [comment]
I downloaded Java 8 (JRE 8).  It doesn't open when I double-click, nor right-click Open.  Could that be because I don't have Service Pack 3?  Same thing happened with a version of Opera I downloaded that seemed to require SP3 (v. 27; v.26 worked).

I found a SP3 file, downloaded it, and installed it several hours ago.  It caused my computer to go into an endless reboot loop.  I researched THAT problem, but didn't find a solution.  I guess I could attempt to download and install SP3 (or even SP4 from MajorGeek).

**Comment by ssoloff** [comment]
```
[0:751] ERROR: No JVM found
[0:751] Download location: https://raw.githubusercontent.com/triplea-game/assets/master/install4j/windows-x86-1.8.0_144.tar.gz
[0:751] Downloading JRE
[1:773] ERROR: InternetOpenUrl failed with error code -2146869244
```

Your log indicates the Win32 `InternetOpenUrl()` function is failing with error code -2146869244 (0x80096004).  The [definition](http://www.lifeasbob.com/code/errorcodes.aspx) of this error code is

```
TRUST_E_CERT_SIGNATURE: The signature of the certificate can not be verified.
```

I suspect that the trusted certificate store for XP SP2 is so old that it doesn't contain the root CA for GitHub, and so can't verify the server certificate when opening the HTTPS URL I pasted above.  Thus, you will not be able to use the bundled Java 8 JRE that is a part of the TripleA installer and must install Java 8 manually.

> I can open that link you posted. In Firefox, it wants me to open or save the file...?

Sorry, I should have explicitly said to use IE.  I think you can open that link in Firefox because Firefox uses its own trusted certificate store rather than the one included in Windows, and that certificate store is probably more up-to-date because (I assume) the version of Firefox you installed is much newer than when XP SP2 was released.  If you use IE, it will use the same trusted certificate store `InternetOpenUrl()` is using, and I would expect it display that security warning I mentioned above.

> I downloaded Java 8 (JRE 8). It doesn't open when I double-click, nor right-click Open. Could that be because I don't have Service Pack 3?

Did you use the "Windows x86 Offline" link when downloading the JRE?  The "Windows x86" link is in a file format (_.tar.gz_) that you wouldn't be able to open on a vanilla XP machine.

I would recommend upgrading to SP3 if possible, as that will give you the best chance to get Java 8 working as expected.  I don't believe there was ever an [XP SP4](https://en.wikipedia.org/wiki/Windows_XP#Service_packs), so I would avoid downloading anything claiming to be so.  (I found [this article](https://www.networkworld.com/article/2602383/microsoft-subnet/windows-xp-service-pack-4-is-a-nice-idea-but-should-be-avoided.html) that explains "SP4" is not an official MSFT product.)

**Comment by RoiEXLab** [comment]
I'll do a quick follow-up:
> I'll copy it below since I'm not sure how to attach a file.

Thanks. That log looks fine and it does indeed confirm the installer fails to download a JRE, so installing it on your own should fix the issue.

> I can open that link you posted. In Firefox, it wants me to open or save the file...?

.@ssoloff Just wanted to let you check you're able to access this site, but I'm almost sure it's the issue he suspects, although your test shows the opposite: Could you try opening the site using Internet Explorer, I'd assume Firefox worksaround this issue by shipping it's own root certificates. There's no need to download anything just make sure you are getting prompted to download something and not getting any error.

> Could that be because I don't have Service Pack 3?

Oh boy... I'd strongly recommend using SP3, although I'd even recommend installing a newer OS.
Not having SP 3 is very likely the issue it seems SP3 is unfortunately no longer available as an official download, so you have to trust one of the various software hosting platforms out there.
But in general: You really shouldn't use XP any longer I know it was a great os, but simply from a security standpoint you're very vulnerable.
I'd recommend installing a lightweight Linux Distribution like [XUbuntu](https://xubuntu.org/) instead of or in addition to XP which will probably still run on your old machine without problems, but is still supported and gets updates on a regular basis.
DOSBox can be installed there as well...
It'll take some time to get used to Linux if you never used it before, but it's really not that difficult as you might think given it has a large online community and you can find guides to almost everything by just googling.
You can even try it before actually installing it after inserting the installation media (DVD-ROM or USB)

**Comment by RoiEXLab** [comment]
Ok, by the time of writing @ssoloff Already answered ^^

**Comment by fasafan** [comment]
Thanks so very much to all!

As I mentioned, the first time I tried to install SP3, the PC went into an endless boot loop.  I figured MajorGeeks' SP4 might be *just* different enough to avoid that problem again.

Linux might be a solid answer, though - especially if I can run DOSBox.  

The funny thing is, whole reason I'm here in the first place is due to an itch to play SSI's Age of Riles.  That led to a desire to play A&A: Iron Blitz - an install that refused to play no matter what I did.  Researching that issue led me to TripleA.  My first play was incredible (although I got my rear handed to me by Turn 3 with the Classic rules/map).  Trying to speed up my machine a bit led to one too many keys being deleted in the registry, and... Reinstall of Windows with only SP2 available.
